### PR TITLE
Tweaks for PSET 10

### PR DESCRIPTION
**Pull request by shrutigarg914**
* fix for REINFORCE autograder to make rewards.shape[1] == max_episode_length
* minor changes in write-up and notation
* addition of a small written question for gradient descent

- - -
This change is [Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/382)

### CODE REVIEW

**Review by nepfaff — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/382)* status: 1 of 4 files reviewed, 4 unresolved discussions (waiting on @shrutigarg914)

___
*[`book/rl/exercises/policy_gradient.ipynb` line 17 at r1](https://reviewable.io/reviews/RussTedrake/manipulation/382#-OB6_anbCCDZPwMto4X5:-OB6_anbCCDZPwMto4X6:b-x5p90g) ([raw file](https://github.com/RussTedrake/manipulation/blob/96116e65c58758e64722c78661d23b911b6db30f/book/rl/exercises/policy_gradient.ipynb#L17)):*
> ```
>     "import numpy as np\n",
>     "import torch\n",
>     "from IPython.display import HTML, display  # , clear_output\n",
> ```

The # looks weird here
___
*[`book/rl/exercises/policy_gradient.ipynb` line 1071 at r1](https://reviewable.io/reviews/RussTedrake/manipulation/382#-OB6_e_F1HbmwMvVvvZw:-OB6_e_F1HbmwMvVvvZx:bw8d5at) ([raw file](https://github.com/RussTedrake/manipulation/blob/96116e65c58758e64722c78661d23b911b6db30f/book/rl/exercises/policy_gradient.ipynb#L1071)):*
> ```
>     "        training_stats[\"avg_episode_length\"].append(mean_episode_lengths)\n",
>     "\n",
>     "    # clear_output()\n",
> ```

Why comment this out?
___
*[`book/rl/exercises/stochastic_optimization.ipynb` line 218 at r1](https://reviewable.io/reviews/RussTedrake/manipulation/382#-OB6a8wR3_mLBiqcILfe:-OB6a8wR3_mLBiqcILff:b-1ekt05) ([raw file](https://github.com/RussTedrake/manipulation/blob/96116e65c58758e64722c78661d23b911b6db30f/book/rl/exercises/stochastic_optimization.ipynb#L218)):*
> ```
>     "## Determinisitc Exact Gradients\n",
>     "\n",
>     "**Problem 11.1.a** [1.5 pts]: Let's first use the standard gradient descent algorithm with exact gradients. Below, you must implement the following simple update function:\n",
> ```

You will need to update the weight in the autograder as well
___
*[`book/rl/exercises/stochastic_optimization.ipynb` line 281 at r1](https://reviewable.io/reviews/RussTedrake/manipulation/382#-OB6aNqf4qkTRE0D8qOt:-OB6aNqg5-ZGrpiaCnfw:b7eommw) ([raw file](https://github.com/RussTedrake/manipulation/blob/96116e65c58758e64722c78661d23b911b6db30f/book/rl/exercises/stochastic_optimization.ipynb#L281)):*
> ```
>     "If you've implemented it correctly, run the cell multiple times to see the behavior of gradient descent from different initial conditions. \n",
>     "\n",
>     "**Problem 11.1.b** [0.5 pts] What do you notice about the behaviour of gradient descent given different starting points? When does it converge or not converge (if ever) to the global minimum? "
> ```

Would you mind providing a reference solution? 

Are we looking for an answer that mentions an attractive region or something similar around the global minimum?

**Comment by shrutigarg914** [comment]
> > Reviewed 1 of 4 files at r1, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/382)_ status: 1 of 4 files reviewed, 4 unresolved discussions (waiting on @shrutigarg914)
> 
> _[`book/rl/exercises/policy_gradient.ipynb` line 17 at r1](https://reviewable.io/reviews/RussTedrake/manipulation/382#-OB6_anbCCDZPwMto4X5:-OB6_anbCCDZPwMto4X6:b-x5p90g) ([raw file](https://github.com/RussTedrake/manipulation/blob/96116e65c58758e64722c78661d23b911b6db30f/book/rl/exercises/policy_gradient.ipynb#L17)):_
> 
> > ```
> >     "import numpy as np\n",
> >     "import torch\n",
> >     "from IPython.display import HTML, display  # , clear_output\n",
> > ```
> 
> The # looks weird here
> 
> _[`book/rl/exercises/policy_gradient.ipynb` line 1071 at r1](https://reviewable.io/reviews/RussTedrake/manipulation/382#-OB6_e_F1HbmwMvVvvZw:-OB6_e_F1HbmwMvVvvZx:bw8d5at) ([raw file](https://github.com/RussTedrake/manipulation/blob/96116e65c58758e64722c78661d23b911b6db30f/book/rl/exercises/policy_gradient.ipynb#L1071)):_
> 
> > ```
> >     "        training_stats[\"avg_episode_length\"].append(mean_episode_lengths)\n",
> >     "\n",
> >     "    # clear_output()\n",
> > ```
> 
> Why comment this out?

According to our notes from last year, it was giving problems with gradescope. So I just took it out as it doesn't seem crucial to getting things to work. I was debating leaving it commented out s.t. students could uncomment it if desired, but I am now just cleanly taking it out.

> _[`book/rl/exercises/stochastic_optimization.ipynb` line 218 at r1](https://reviewable.io/reviews/RussTedrake/manipulation/382#-OB6a8wR3_mLBiqcILfe:-OB6a8wR3_mLBiqcILff:b-1ekt05) ([raw file](https://github.com/RussTedrake/manipulation/blob/96116e65c58758e64722c78661d23b911b6db30f/book/rl/exercises/stochastic_optimization.ipynb#L218)):_
> 
> > ```
> >     "## Determinisitc Exact Gradients\n",
> >     "\n",
> >     "**Problem 11.1.a** [1.5 pts]: Let's first use the standard gradient descent algorithm with exact gradients. Below, you must implement the following simple update function:\n",
> > ```
> 
> You will need to update the weight in the autograder as well

Done!

> _[`book/rl/exercises/stochastic_optimization.ipynb` line 281 at r1](https://reviewable.io/reviews/RussTedrake/manipulation/382#-OB6aNqf4qkTRE0D8qOt:-OB6aNqg5-ZGrpiaCnfw:b7eommw) ([raw file](https://github.com/RussTedrake/manipulation/blob/96116e65c58758e64722c78661d23b911b6db30f/book/rl/exercises/stochastic_optimization.ipynb#L281)):_
> 
> > ```
> >     "If you've implemented it correctly, run the cell multiple times to see the behavior of gradient descent from different initial conditions. \n",
> >     "\n",
> >     "**Problem 11.1.b** [0.5 pts] What do you notice about the behaviour of gradient descent given different starting points? When does it converge or not converge (if ever) to the global minimum? "
> > ```
> 
> Would you mind providing a reference solution? Are we looking for an answer that mentions an attractive region or something similar around the global minimum?

Done! Just  looking for an understanding that that we converge to local minima if we start in their region of attractions aka that we don't always converge to global minimum.

**Review by nepfaff — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RussTedrake/manipulation/382)* status: all files reviewed, 4 unresolved discussions (waiting on @shrutigarg914)

___
*[`book/rl/exercises/policy_gradient.ipynb` line 1071 at r1](https://reviewable.io/reviews/RussTedrake/manipulation/382#-OB6_e_F1HbmwMvVvvZw:-OB7UK4cDTRY5ly108JK:bil7lsi) ([raw file](https://github.com/RussTedrake/manipulation/blob/96116e65c58758e64722c78661d23b911b6db30f/book/rl/exercises/policy_gradient.ipynb#L1071)):*
<details><summary><i>Previously, shrutigarg914 (Shruti Garg) wrote…</i></summary><blockquote>

According to our notes from last year, it was giving problems with gradescope. So I just took it out as it doesn't seem crucial to getting things to work. I was debating leaving it commented out s.t. students could uncomment it if desired, but I am now just cleanly taking it out.
</blockquote></details>

Sounds good. Thanks!
___
*[`book/rl/exercises/stochastic_optimization.ipynb` line 218 at r1](https://reviewable.io/reviews/RussTedrake/manipulation/382#-OB6a8wR3_mLBiqcILfe:-OB7UNI704bM5U0-f2PG:b-ucfknh) ([raw file](https://github.com/RussTedrake/manipulation/blob/96116e65c58758e64722c78661d23b911b6db30f/book/rl/exercises/stochastic_optimization.ipynb#L218)):*
<details><summary><i>Previously, shrutigarg914 (Shruti Garg) wrote…</i></summary><blockquote>

Done!
</blockquote></details>

Thanks!
___
*[`book/rl/exercises/stochastic_optimization.ipynb` line 281 at r1](https://reviewable.io/reviews/RussTedrake/manipulation/382#-OB6aNqf4qkTRE0D8qOt:-OB7Uco2EPqHfIVJloN0:b-ucfknh) ([raw file](https://github.com/RussTedrake/manipulation/blob/96116e65c58758e64722c78661d23b911b6db30f/book/rl/exercises/stochastic_optimization.ipynb#L281)):*
<details><summary><i>Previously, shrutigarg914 (Shruti Garg) wrote…</i></summary><blockquote>

Done! Just  looking for an understanding that that we converge to local minima if we start in their region of attractions aka that we don't always converge to global minimum.
</blockquote></details>

Thanks!
